### PR TITLE
Soulkeys

### PR DESCRIPTION
**Pull request by CryptoGraffe**
### <samp>🤖 Generated by Copilot at 6e838c9</samp>

This pull request implements the SoulKeys feature, which allows users to mint and transfer NFTs representing shares of other users' souls. It adds a new contract, a new action, and new tables to the Hardhat and Hasura layers, respectively. It also updates the configuration, deployment, management, and test scripts for the SoulKeys contract. 

### <samp>🤖 Generated by Copilot at 6e838c9</samp>

> _`syncKeys` action_
> _SoulKeys contract on chain_
> _Autumn of the soul_

### <samp>🤖 Generated by Copilot at 6e838c9</samp>

*  Implement the SoulKeys feature, which allows users to buy and sell shares of other users' souls ([link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-0eb92345899bdcb82868ad7ed09d43c78c75a832429c714563295a8d0db369a8R1-R218), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-a6a476461d778e69905cd6d1cab214af3b0f749568bce39067fdde541f73a24eR38), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-a6a476461d778e69905cd6d1cab214af3b0f749568bce39067fdde541f73a24eR86), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-980b9e1b362aa0d0dbd9ac63c605b4fc810dd8876c5f15290484377a8f9ba007R244-R246), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-e46a68635b688f3bc70fa7920f0d043beb4bf16e4d9ff2b0fb7e74caa48bf38dL1-R40), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-164e2e324cf362f6b843e2c142bb947c8d5b84ce92f29998a00eef54bcf8877fL18-R19), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-164e2e324cf362f6b843e2c142bb947c8d5b84ce92f29998a00eef54bcf8877fR247-R249), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-164e2e324cf362f6b843e2c142bb947c8d5b84ce92f29998a00eef54bcf8877fL278-R285), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-5435b181502f5f5c7e7c8bc1f0733489f71553e0d4660737cce1c3568d150970R13), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-a494539efa0672685daa279c2b93635d0bfa0a2264689910a8122ed0c4a1347bR1-R33), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-8054e2185283d4b075b349962edd5d35c7fde0112d3ebb7e683706f8e9bcfa2aR1-R60), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-229ea93b29cd5c3b63805aa068bd499629074aef2fbd70e6632dc390f65b0659R1-R161), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-b734c8ed9cd439ea59975ad4befcf67d6722e79bca44bbf0ea2105d2999f0b02L15-R20), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-b734c8ed9cd439ea59975ad4befcf67d6722e79bca44bbf0ea2105d2999f0b02R27-R28), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-b734c8ed9cd439ea59975ad4befcf67d6722e79bca44bbf0ea2105d2999f0b02R46), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-b734c8ed9cd439ea59975ad4befcf67d6722e79bca44bbf0ea2105d2999f0b02R98-R101), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-589f402dbb6eb11ea9c28e067ac01d020e1b8f8f087f45d59435eaae684975c3R130-R133), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-49e7aa4e1112b93bcdbd36419aeb0b64fb23ea532e80c9a0f5ff1c5a585763f9R306-R315), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-3837acf731f554c94063419d1b6e3b165007c982ac40c9d93b10416b78fd9532R23-R41), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-fff82f2e4d3da1dcf7e90bfe43e3be3a7f9172b1ae7a16e4c9f3fd613d93bc87R1-R43), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-5c75f0511ae1a1cf2c55e9db0801b41bc422dee8dfe3653bd06277115591cd55R1-R38))
  * Add a handler function for the `syncKeys` action, which syncs the key transactions and holders from the SoulKeys contract on the blockchain to the Hasura database (`./api/hasura/actions/_handlers/syncKeys.ts`, [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-0eb92345899bdcb82868ad7ed09d43c78c75a832429c714563295a8d0db369a8R1-R218))
  * Import and register the `syncKeys` handler function in the `actionManager.ts` file, which routes the action requests from Hasura to the appropriate handlers (`./api/hasura/actions/actionManager.ts`, [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-a6a476461d778e69905cd6d1cab214af3b0f749568bce39067fdde541f73a24eR38), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-a6a476461d778e69905cd6d1cab214af3b0f749568bce39067fdde541f73a24eR86))
  * Add the address of the deployed SoulKeys contract to the `hardhat/deploymentInfo.json` file, which contains information about the deployed contracts on different networks ([link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-980b9e1b362aa0d0dbd9ac63c605b4fc810dd8876c5f15290484377a8f9ba007R244-R246))
  * Add the Solidity compiler input for the SoulKeys contract to the `hardhat/deployments/optimismGoerli/solcInputs/bb81b2312b02609cb88e7e560a0ed6dc.json` file, which stores the compiler input for verification and debugging purposes ([link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-e46a68635b688f3bc70fa7920f0d043beb4bf16e4d9ff2b0fb7e74caa48bf38dL1-R40))
  * Update the Hardhat configuration file for the SoulKeys feature, by sorting the imports alphabetically, adding the `feeDestination` field to the `config` object, and modifying the `optimismGoerli` network configuration (`./hardhat/hardhat.config.ts`, [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-164e2e324cf362f6b843e2c142bb947c8d5b84ce92f29998a00eef54bcf8877fL18-R19), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-164e2e324cf362f6b843e2c142bb947c8d5b84ce92f29998a00eef54bcf8877fR247-R249), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-164e2e324cf362f6b843e2c142bb947c8d5b84ce92f29998a00eef54bcf8877fL278-R285))
  * Add a new script `test` to the `hardhat/package.json` file, which runs the Hardhat test command ([link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-5435b181502f5f5c7e7c8bc1f0733489f71553e0d4660737cce1c3568d150970R13))
  * Add a deployment script for the SoulKeys contract, which deploys the contract to the Optimism Goerli network using Hardhat (`./hardhat/scripts/deploy/04-soulkeys/00-soulkeys-deploy.ts`, [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-a494539efa0672685daa279c2b93635d0bfa0a2264689910a8122ed0c4a1347bR1-R33))
  * Add a management script for the SoulKeys contract, which sets the fee destination, protocol fee percent, and subject fee percent for the contract using Hardhat (`./hardhat/scripts/manage/soulkeys_setup.ts`, [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-8054e2185283d4b075b349962edd5d35c7fde0112d3ebb7e683706f8e9bcfa2aR1-R60))
  * Add a test suite for the SoulKeys contract, which tests the functionality and behavior of the contract using Hardhat (`./hardhat/test/02-soulkeys/00-soulkeys.ts`, [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-229ea93b29cd5c3b63805aa068bd499629074aef2fbd70e6632dc390f65b0659R1-R161))
  * Update the deployment utility file for the SoulKeys feature, by sorting the imports alphabetically, importing the `SoulKeys` and `SoulKeys__factory` types, adding the `soulKeys` field to the `DeployedContracts` type, and assigning the `soulKeys` field to the deployed SoulKeys contract (`./hardhat/test/utils/deployment.ts`, [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-b734c8ed9cd439ea59975ad4befcf67d6722e79bca44bbf0ea2105d2999f0b02L15-R20), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-b734c8ed9cd439ea59975ad4befcf67d6722e79bca44bbf0ea2105d2999f0b02R27-R28), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-b734c8ed9cd439ea59975ad4befcf67d6722e79bca44bbf0ea2105d2999f0b02R46), [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-b734c8ed9cd439ea59975ad4befcf67d6722e79bca44bbf0ea2105d2999f0b02R98-R101))
  * Define the GraphQL schema for the `syncKeys` action, by adding a new field `syncKeys` to the `Mutation` type in the `hasura/metadata/actions.graphql` file ([link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-589f402dbb6eb11ea9c28e067ac01d020e1b8f8f087f45d59435eaae684975c3R130-R133))
  * Configure the `syncKeys` action with its name, definition, handler, and permissions, by adding a new entry for the action in the `hasura/metadata/actions.yaml` file ([link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-49e7aa4e1112b93bcdbd36419aeb0b64fb23ea532e80c9a0f5ff1c5a585763f9R306-R315))
  * Update the Hasura database schema for the SoulKeys feature, by adding two new array relationships `held_keys` and `key_holders` to the `cosouls` table, and creating the `key_holders` and `key_tx` tables with their object relationships and select permissions (`./hasura/metadata/databases/default/tables/public_cosouls.yaml`, [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-3837acf731f554c94063419d1b6e3b165007c982ac40c9d93b10416b78fd9532R23-R41), `./hasura/metadata/databases/default/tables/public_key_holders.yaml`, [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-fff82f2e4d3da1dcf7e90bfe43e3be3a7f9172b1ae7a16e4c9f3fd613d93bc87R1-R43), `./hasura/metadata/databases/default/tables/public_key_tx.yaml`, [link](https://github.com/coordinape/coordinape/pull/2362/files?diff=unified&w=0#diff-5c75f0511ae1a1cf2c55e9db0801b41bc422dee8dfe3653bd06277115591cd55R1-R38))

### CODE REVIEW

File: hardhat/test/02-soulkeys/00-soulkeys.ts [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+      0
+    );
+
+    const price = await soulKeys.getPrice(supply, 1);
+    const fees = price.mul(2).div(20 /* five percent */).toString();
+    expect(price.toString()).to.eq('62500000000000');
+
+    const buyPrice = await soulKeys.getBuyPriceAfterFee(subject.address, 1);
+    const expectedBuyPrice = price.add(fees);
+    console.log('buyPrice', ethers.utils.formatEther(buyPrice));
```
Comment (crabsinger) [inline hardhat/test/02-soulkeys/00-soulkeys.ts:72]: remove console logs from tests

File: hardhat/test/02-soulkeys/00-soulkeys.ts [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        .add(sellPrice)
+        .sub(txr.gasUsed.mul(txr.effectiveGasPrice))
+        .toString()
+    );
+
+    const tradeSig =
+      'Trade(address,address,bool,uint256,uint256,uint256,uint256,uint256)';
+    const tradeTopic: string = ethers.utils.id(tradeSig);
+
+    const currentBlock = await ethers.provider.getBlockNumber();
```
Comment (crabsinger) [inline hardhat/test/02-soulkeys/00-soulkeys.ts:138]: remove - no expectation on these.

we can do a ```
    await expect(contract.setSlot(0, 6969, tokenId))
      .to.emit(cosoul, 'MetadataUpdate')
      .withArgs(tokenId);
```

 like in cosoul test

File: api/hasura/actions/_handlers/syncKeys.ts [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    return errorResponse(res, e);
+  }
+}
+
+type InsertOrUpdateHolder = Pick<
+  Required<ValueTypes['key_holders_insert_input']>,
+  'address' | 'subject' | 'amount'
+> & { amount: number };
+
+// this goes over ever trade the address has ever done. could me more efficient
```
Comment (crabsinger) [inline api/hasura/actions/_handlers/syncKeys.ts:138]: `s/could me/could be`

File: hardhat/scripts/manage/soulkeys_setup.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+// const WEI = 0.000000001;
+const FIVE_PERCENT_IN_WEI = BigNumber.from('50000000000000000'); // (1.0 / WEI) * 0.05;
+
+async function main() {
+  try {
+    const { deployer, feeDestination } = await hre.getNamedAccounts();
+
+    const deployed_soulkeys = await hre.deployments.get('SoulKeys');
+
+    // use pgiveSyncer account - which has perms for setting mintFee
```
Comment (crabsinger) [inline hardhat/scripts/manage/soulkeys_setup.ts:17]: this is deployer, not pgiveSyncer?

File: src/features/soulkeys/CoSoulChainGate.tsx [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  actionName,
+  children,
+}) => {
+  const { library, chainId, account } = useWeb3React();
+  const contracts = useCoSoulContracts();
+  const { showError } = useToast();
+
+  const onCorrectChain = chainId === Number(chain.chainId);
+
+  const safeSwitchToCorrectChain = async () => {
```
Comment (crabsinger) [inline src/features/soulkeys/CoSoulChainGate.tsx:29]: could reuse this `safe` switch function

File: src/features/soulkeys/useSoulKeys.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  subject: string;
+}) => {
+  const [balance, setBalance] = useState<number | null>(null);
+
+  const queryClient = useQueryClient();
+  const refresh = () => {
+    contracts.soulKeys
+      .sharesBalance(subject, address)
+      .then(b => setBalance(b.toNumber()));
+    setTimeout(() => {
```
Comment (crabsinger) [inline src/features/soulkeys/useSoulKeys.ts:23]: seems fishy...

Comment (CryptoGraffe) [reply src/features/soulkeys/useSoulKeys.ts:23]: this is due to lack of webhooks and stuff taking long time, not sure what else to do

File: hardhat/test/02-soulkeys/00-soulkeys.ts (hunk 1 above)
Comment (CryptoGraffe) [reply hardhat/test/02-soulkeys/00-soulkeys.ts]: done

File: api/hasura/actions/_handlers/syncKeys.ts (hunk 3 above)
Comment (CryptoGraffe) [reply api/hasura/actions/_handlers/syncKeys.ts]: done

File: hardhat/scripts/manage/soulkeys_setup.ts (hunk 4 above)
Comment (CryptoGraffe) [reply hardhat/scripts/manage/soulkeys_setup.ts:17]: done

File: src/features/soulkeys/useSoulKeys.ts (hunk 6 above)
Comment (CryptoGraffe) [reply src/features/soulkeys/useSoulKeys.ts:23]: yep but its ok for now until webhook

**Review by crabsinger — APPROVED** [review]

**Pull request merged**